### PR TITLE
[B+C] Add Wake and Sleep API. Adds BUKKIT-3155

### PR DESCRIPTION
**Pull request by JadenKorr**
**The Issue:**
There is no way of waking players up, neither is there a way to put them in a bed.

**Justification for this PR:**
This feature was requested in the #bukkitdev IRC channel by a few plugin developers.

**PR Breakdown:**
Adds three new API calls to the Player Entity. 2 for sleeping, 1 for waking. Als adds a new Enum to allow checking if putting the player to sleep succeeded.

**Testing Results and Materials:**
This PR can be tested by the following test plugin: https://github.com/JadenKorr/SleepWakeTestPlugin

**Relevant PR(s):**
CB-953 - https://github.com/Bukkit/CraftBukkit/pull/953

**JIRA Ticket:**
BUKKIT-3155 - https://bukkit.atlassian.net/browse/BUKKIT-3155

**Pull Request Check List:**
_General_:
- [ ] Fits Bukkit's Goals
- [ ] Leaky Ticket Ready
- [ ] Code Meets Requirements
- [ ] Code is Documented
- [ ] Code Addresses Leaky Ticket
- [ ] Followed Pull Request Format
- [ ] Tested Code
- [ ] Included Test Material and Source

### CODE REVIEW

File: src/main/java/org/bukkit/entity/Player.java [hunk 1]
```diff
@@ -495,6 +495,28 @@
     public void setBedSpawnLocation(Location location, boolean force);
 
     /**
+     * Attempts to make the Player sleep in the bed at the given location
+     *
+     * @param location where the bed the bed is at
```
Comment (jrtc27) [inline src/main/java/org/bukkit/entity/Player.java]: "where the bed the bed is at" - maybe something like `The location of the bed"?

**Comment by jrtc27** [comment]
Can you fix your capitalisation? Please? :D

**Comment by jrtc27** [comment]
Oh you might want to edit the description of this request so that it says "CraftBukkit PR" not "Bukkit PR".

File: src/main/java/org/bukkit/BedResult.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        return BY_ID.get(id);
+    }
+
+    static {
+        for (BedResult result : values()) {
+            BY_ID.put(result.id, result);
+        }
+    }
+}
\ No newline at end of file
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/BedResult.java]: Needs a newline.

**Comment by gmcferrin** [comment]
@JadenKorr The Javadocs for this are pretty ugly. Could you try to rework them a bit? Also, the naming convention used for the enum is a bit awkward.

**Comment by turt2live** [comment]
@JadenKorr If you are still interested in this please update it to the latest code base. The PR will also need to match the CONTRIBUTING.md guidelines and the code updated to accommodate the concerns above. 

Once these items are completed then the review process shall continue.

**Comment by JadenKorr** [comment]
I forgot that I still had this PR open. I just rebased it on Bukkit/Bukkit and applied my changes to it, I have however changed up the wording in the Javadoc and renamed the enum and its content to be more concise.

I've also updated the title and PR text to follow the guidelines.

CB PR is now updated

File: src/main/java/org/bukkit/SleepingResult.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package org.bukkit;
+
+public enum SleepingResult {
```
Comment (turt2live) [inline src/main/java/org/bukkit/SleepingResult.java:3]: Everything in this enum needs 2 things:
1) Renaming to be more appropriate
2) JavaDocs.

File: src/main/java/org/bukkit/entity/Player.java [hunk 4]
```diff
@@ -547,6 +548,30 @@
     public void setBedSpawnLocation(Location location, boolean force);
 
     /**
+    * Attempts to put the player to sleep in the bed at given Location.
+    *
+    * @param bed the location of the bed.
+    * @return OK, if the player is put to sleep. Else the reason why not.
```
Comment (turt2live) [inline src/main/java/org/bukkit/entity/Player.java:554]: "OK"? Unless I actually look at stuff I assume this is a true/false value. Either link the entire enum, or reword this.

All subsequent JavaDocs will need to be edited as well

File: src/main/java/org/bukkit/SleepingResult.java (hunk 3 above)
Comment (JadenKorr) [inline src/main/java/org/bukkit/SleepingResult.java:3]: In NMS its called BedResult. I could name it like that, I just thought SleepingResult was more appropriate/self-explanatory. And I guess only the Enum it self needs a JavaDoc, the Enum values are self-explanatory

Comment (turt2live) [inline src/main/java/org/bukkit/SleepingResult.java:3]: Bukkit does not need to follow NMS to a tee. 

This enum does not exactly make sense as there is no description about what SleepingResult.OK (for example) actually means.

**Comment by EdGruberman** [comment]
Please add new commits instead of forcibly updating existing commits so we can follow along to see what has changed over time and get a history of changes.

You might consider renaming `SleepingResult` to `BedResult`.  I've always distinguished between the action of entering and exiting a bed versus an actual sleep cycle (night time changed to day time) completing.  It would be good to ensure the distinction continues.

**Pull request closed without merging**